File: Data-Structures-and-Algorithms/0015-3Sum/threeSum.py
class Solution(object):
    def threeSum(self, nums):
        # sorting so that anything above 0 is discarded (impossible to make a zero out of only positive numbers)
        nums.sort()   #O(N*logN)
        res = []
       
        for i in range(len(nums)):
            # if the number is above 0, we break this outer for loop
            if nums[i] > 0:
                break
            # if the negative is unique from previous element value, we run the twoSum function that finds the complement
            if i == 0 or nums[i] != nums[i-1]:
                self.twoSum(nums, i, res)
        return res

    def twoSum(self, nums, i, res): 
        # create a set, which records numbers to the right of nums[i]
        seen = set()
        # start with i + 1, to find more pairs with -nums[i]
        j = i + 1
        while j < len(nums):
            # if comp is in seen set, then we know the comp, the nums[i], and the nums[j] is a triplet.
            comp = -(nums[i]+nums[j])
            if comp in seen:
                res.append([nums[i], nums[j], comp])
                # if the j+1 element is the same as the j element, we skip
                while j + 1 < len(nums) and nums[j] == nums[j + 1]:
                    j += 1
            # this way, the numbers won't be repeated because it is a set, and the returns won't repeat itself
            seen.add(nums[j])
            j += 1

nums = [-1,0,1,2,-1,-4]
print(Solution().threeSum(nums))

# O(N^2) since it has two full loops (almost) within one another
# O(N) stores hashset of size N
# Array of numbers with negative and positive values, find the threeSum that is zero

File: Data-Structures-and-Algorithms/0015-3Sum/threeSum.js
const threeSum = function(nums) {
    let ans = [];
    // impossible to make a 0 out of less than 3 numbers per requested
    if (nums.length < 3) {
        return []
    }
    // Sort the array in increasing (ascending) order.
    nums.sort((a,b) => a - b)
    // Outer for loop
    for (let i = 0; i < nums.length - 2; i++) {
        // Skip numbers that are equal to their immediate predecessors
        if (i > 0 && nums[i] == nums[i-1]) {
            continue
        }
        // break the loop is the number is greater or equal to zero because you cant make zero out of them
        if (nums[i] >= 0) {
            break
        }
        // With nums[i] in hand, find a twoSum using two pointers approach from two sides of the rest of the array
        let left = i+1;
        let right = nums.length-1;

        while (left < right) {
            let cursum = nums[i] + nums[left] + nums[right]
            if (cursum < 0) {
                left ++;
            } else if (cursum > 0) {
                right --;
            } else if (cursum == 0) { 
                ans.push([nums[i], nums[left], nums[right]])
                while(nums[left]===nums[left+1]){left++}
                while(nums[right]===nums[right-1]){right--}
                left++;
                right--;
            }
        }
    }

    return ans
}

nums = [-1,0,1,2,-1,-4]
console.log(threeSum(nums))

// # O(N^2 + NlogN) ~ O(N^2) since it has two loops (almost) within one another
// O(N) for the sorting function
// Array of numbers with negative and positive values, find the threeSum that is zero